### PR TITLE
Improve GPU training batching and benchmarks

### PR DESCRIPTION
**Pull request by zazabap**
## Summary
- pre-stack image batches before optimizer iterations so batched GPU losses do not rebuild identical batch tensors on each AD/loss evaluation
- add pre-stacked batch loss APIs and tests for L1/L2/MSE
- update `examples/optimizer_benchmark.jl` to use QuickDraw and compare customized Riemannian GD with Manopt.jl Riemannian gradient descent using equal full-batch step counts
- set the benchmark GPU selection to device 1
- update the benchmark submodule pointer to include the GPU profiling tool from zazabap/ParametricDFT-Benchmarks.jl#2

## Testing
- `julia --project=. -e 'using ParametricDFT, Test, Random, OMEinsum, Zygote; include("test/loss_tests.jl")'`\n- `julia --project=. -e 'using ParametricDFT, Test, Random, LinearAlgebra; include("test/training_tests.jl")'`\n- `julia --project=. -e 'Meta.parseall(read("examples/optimizer_benchmark.jl", String)); Meta.parseall(read("examples/benchmark/profile_gpu_training.jl", String)); println("parse ok")'`\n\n## Notes\n- GPU 1 exists on this machine (`nvidia-smi -L` shows two RTX 3090s), but CUDA returned `ERROR_NOT_PERMITTED` when selecting device 1, and `CUDA_VISIBLE_DEVICES=1` reported no CUDA device. I could not complete the live GPU smoke run here.\n- `docs/drafts/` was left untracked.\n

### CODE REVIEW

**Comment by zazabap** [comment]
## Redundancy Cleanup Summary

### Overview

This batch removes duplicated/dead code from the four core modules (`loss.jl`, `manifolds.jl`, `optimizers.jl`, `training.jl`) while keeping the public API unchanged (except removing the useless `L2Norm`). All 1354 tests pass. Benchmark results are identical before and after.

### Line Count

| File | Before | After | Delta |
|------|--------|-------|-------|
| `src/loss.jl` | 330 | 283 | **-47** |
| `src/manifolds.jl` | 228 | 231 | +3 |
| `src/optimizers.jl` | 359 | 389 | +30 |
| `src/training.jl` | 593 | 466 | **-127** |
| **Total** | **1510** | **1369** | **-141** |

### Changes by File

#### `src/loss.jl` (330 → 283 lines)

- **Removed `L2Norm`** — struct, dispatch, batched functions, wrappers. L2 norm is constant under unitary transforms (Parseval's theorem), gradient is always zero.
- **Added `_ensure_tuple` helper** — replaces 7 one-liner Vector-to-Tuple wrapper methods with a single conversion point.
- **Collapsed `loss_function` overloads** — from 8 methods down to 3 (single image, batch, stacked array). Extracted `_loss_function_batched` private helper.
- **Added explicit error branch** — `else error("unsupported loss type")` instead of unsafe fallthrough.

#### `src/manifolds.jl` (228 → 231 lines)

- **Extracted `_make_identity_batch(T, d, n)`** — replaces 3 inlined identity-batch construction loops (in `retract`, GD setup, Adam setup) with a single helper.

#### `src/optimizers.jl` (359 → 389 lines)

- **Added `OptimizationState{ET,RT}`** — bundles shared loop state (manifold groups, point batches, gradient buffers, identity cache).
- **Added `_common_setup(tensors)`** — builds `OptimizationState`, classifies manifolds, stacks tensors, pre-allocates buffers.
- **Added `_optimization_loop`** — single shared iteration loop (unstack → gradients → project → convergence check → update → loss trace).
- **Added per-optimizer dispatch**:
  - `_init_optimizer_state(::RiemannianGD)` → `nothing`; `_init_optimizer_state(::RiemannianAdam)` → moment buffers
  - `_update_step!(::RiemannianGD)` → Armijo line search; `_update_step!(::RiemannianAdam)` → moment update + retract
- **Replaced two `optimize!` methods** with a single `optimize!(::AbstractRiemannianOptimizer, ...)`.

#### `src/training.jl` (593 → 466 lines)

- **Added per-type dispatch interface**:
  - `_init_circuit(::Type{B}, m, n; kwargs...)` — circuit initialization per basis type
  - `_build_basis(::Type{B}, m, n, tensors, oc, ic; kwargs...)` — basis construction per basis type
  - `_basis_name(::Type{B})` — string name per basis type
- **Replaced 4 `train_basis` methods** (~200 lines) with 1 generic `train_basis(::Type{B}, ...) where B <: AbstractSparseBasis` (~40 lines) + dispatch table.
- **Batched validation loss** — `_compute_validation_loss` now uses batched einsum codes when available (consistent with training path).

### Benchmark Verification

Optimizer benchmark (`examples/optimizer_benchmark.jl`) produces identical results:

| Config | Before | After |
|--------|--------|-------|
| **QuickDraw (32×32)** | | |
| Manopt-GD | 44.57s, loss=109.29 | 43.39s, loss=109.29 |
| PDFT-GD (cpu) | 2.75s, loss=28.01 | 2.68s, loss=28.01 |
| PDFT-GD (gpu) | 1.64s, loss=28.16 | 1.58s, loss=28.16 |
| **DIV2K (512×512)** | | |
| PDFT-GD (cpu) | 107.03s, loss=658.97 | 101.53s, loss=658.97 |
| PDFT-GD (gpu) | 18.44s, loss=661.01 | 18.54s, loss=661.01 |
| PDFT-Adam (cpu) | 60.85s, loss=658.13 | 59.62s, loss=658.13 |
| PDFT-Adam (gpu) | 14.39s, loss=660.79 | 14.37s, loss=660.79 |

Loss values are identical (deterministic). Timings within normal variance.

### Commits

```
833c7f3 docs: update optimizer_benchmark.jl with QuickDraw + DIV2K dual-profile support
06c79ec perf: use batched einsum for validation loss computation
f053252 refactor: consolidate four train_basis methods into generic + per-type dispatch
a792f9a fix: replace fragile type introspection with accessors, simplify _update_step! return
3281456 refactor: extract shared _optimization_loop with per-optimizer _update_step! dispatch
1f33a49 refactor: extract _make_identity_batch helper from retract
2170403 refactor: replace Vector-to-Tuple wrappers with _ensure_tuple helper
643dafc fix: explicit loss type dispatch + unwrap single-element test loop
5495b89 refactor: remove L2Norm (gradient is always zero under unitary transforms)
```
